packages/core/src/StateNode.ts

```diff
@@ -972,23 +972,24 @@ class StateNode<
   }
 
   private getExternalReentryNodes(
-    childStateNode: StateNode<TContext, any, TEvent, any, any, any>
+    targetNode: StateNode<TContext, any, TEvent, any, any, any>
   ): Array<StateNode<TContext, any, TEvent, any, any, any>> {
     const nodes: Array<StateNode<TContext, any, TEvent, any, any, any>> = [];
-    let marker:
-      | StateNode<TContext, any, TEvent, any, any, any>
-      | undefined = childStateNode;
+    let [marker, other]: [
+      StateNode<TContext, any, TEvent, any, any, any> | undefined,
+      StateNode<TContext, any, TEvent, any, any, any>
+    ] = targetNode.order > this.order ? [targetNode, this] : [this, targetNode];
```

Comment:
somewhat neat trick with comparing those orders here:
- `targetNode.order > this.order` means that `targetNode` is either a descendant of `this` OR it's somewhere outside of it, within the latter siblings of one of its ancestors
- `targetNode.order < this.order` means that `targetNode` is either the ancestor of `this` OR it's somewhere outside of it, within the former siblings of one of its ancestors

Since this function returns `[]` when those two do not form an ancestor<->descendant relationship we can safely rely on this as we are not even concerned here with finding a common ancestor of the two, we only want to find out if one is the ancestor of the other one and collect all nodes in between if that's the case (including the detected ancestor node).

Reply:
Very clever! To prevent any edge-cases, we should ensure that `order` is not configurable (I think it technically is right now)

Reply:
I don't think there is currently any risk here because `order` is a local variable initialized to a static `0` and only assigned within the `dfs` call here:
https://github.com/statelyai/xstate/blob/e35493f59d277ca57f0982417d5ba3bca0a352ed/packages/core/src/StateNode.ts#L374-L385